training_methods/automl_training_method.py

```diff
@@ -0,0 +1,440 @@
+                write_disposition="WRITE_TRUNCATE",
+            )
+
+            job = client.load_table_from_dataframe(
+                dataframe=evaluation_metrics_df,
+                destination=table_id,
+                job_config=job_config,
+            )
+            # Wait for the job to complete.
+            _ = job.result()
+
+            return str(job.destination)
+        else:
+            raise ValueError(
+                f"Model evaluation data does not exist for model {model_name}!"
+            )
+
+    def _predict(self, model_name: str, bigquery_source: str) -> str:
+        """This function runs the batch prediction job
+
+        Args:
+            model_name (str): The model used for batch prediciton.
+            bigquery_source (str): The BigQuery source URI for batch prediction.
+
+        Returns:
+            str: The table id of batch prediction results.
+        """
+
+        client = bigquery.Client()
+        project_id = client.project
+
+        model = aiplatform.Model(model_name=model_name)
+
+        job = model.batch_predict(
+            job_display_name=f"automl_forecasting_{utils.generate_uuid()}",
+            bigquery_source=bigquery_source,
+            instances_format="bigquery",
+            bigquery_destination_prefix=f"bq://{project_id}",
+            predictions_format="bigquery",
+            generate_explanation=True,
+            sync=True,
+        )
+
+        output_dataset = job.output_info.bigquery_output_dataset
+        output_dataset = output_dataset.replace("bq://", "")
+        output_table = job.output_info.bigquery_output_table
+        bq_output_table_id = f"{output_dataset}.{output_table}"
+
+        return bq_output_table_id
+
+    def _prepare_test_dataset(
+        self,
+        context_window: int,
+        forecast_horizon: int,
+        model_parameters: Dict[str, Any],
+        bigquery_source: str,
+    ) -> str:
+        """This function does the prerocessing job on the test data and
+        saves the result in a bigquery table
+
+        Args:
+            context_window (int): Sets how far back the model looks during training
+              (and for forecasts).
+            forecast_horizon (int): Determines how far into the future the model
+              forecasts the target value for each row of prediction data.
+            model_parameters (Dict[str, Any]): The model training parameters.
+            bigquery_source (str): BigQuery uri of the table to be processed in
+              format bq://project-id.dataset-id.table-id
+
+        Returns:
+            str: BigQuery uri of the destination table where the preprocess data is
+              saved to in format bq://project-id.dataset-id.table-id
+        """
+
+        time_column = model_parameters.get(TIME_COLUMN_PARAMETER)
+        target_column = model_parameters.get(TARGET_COLUMN_PARAMETER)
+
+        # Download dataset from BigQuery
```

Comment:
Maybe add validation here too.

Reply:
Done!

Reply:
I think we're repeating these validations in the code! There might be a better way to do all those checks in one place :-? One way could be creating `REQUIRED_PARAMETERS` list and validate that base on that list before anything when a training method is called! We could do the validation in __init__ methods. WDYT?

Reply:
I think either is fine. 
Don't think you can do it in `init` as the validation happens when `train` is called.